packages/flutter_tools/lib/src/localizations/message_parser.dart

```diff
@@ -0,0 +1,519 @@
+        //   startIndex = match.end;
+        //   continue;
+        // }
+        match = unescapedString.matchAsPrefix(message, startIndex);
+        if (match != null) {
+          tokens.add(Node.string(startIndex, match.group(0)!));
+          startIndex = match.end;
+          continue;
+        }
+        match = brace.matchAsPrefix(message, startIndex);
+        if (match != null) {
+          tokens.add(Node.brace(startIndex, match.group(0)!));
+          isString = false;
+          startIndex = match.end;
+          continue;
+        }
+        // Theoretically, we only reach this point because of unmatched single quotes because
+        // 1. If it begins with single quotes, then we match the longest string contained in single quotes.
+        // 2. If it begins with braces, then we match those braces.
+        // 3. Else the first character is neither single quote or brace so it is matched by RegExp "unescapedString"
+        throw L10nException('''
+ICU Lexing Error: Unmatched single quotes.
+$message
+${indentForError(startIndex)}''');
+      } else {
+        RegExp matcher;
+        ST? matchedType;
+
+        // Try to match tokens until we succeed
+        for (matchedType in matchers.keys) {
+          matcher = matchers[matchedType]!;
+          match = matcher.matchAsPrefix(message, startIndex);
+          if (match != null) {
+            break;
+          }
+        }
+
+        if (match == null) {
+          match = brace.matchAsPrefix(message, startIndex);
+          if (match != null) {
+            tokens.add(Node.brace(startIndex, match.group(0)!));
+            isString = true;
+            startIndex = match.end;
+            continue;
+          }
+          // This should only happen when there are special characters we are unable to match.
+          throw L10nException('''
+ICU Lexing Error: Unexpected character.
+$message
+${indentForError(startIndex)}''');
+        } else if (matchedType == ST.empty) {
+          // Do not add whitespace as a token.
+          startIndex = match.end;
+          continue;
+        }else {
+          tokens.add(Node(matchedType!, startIndex, value: match.group(0)));
+          startIndex = match.end;
+          continue;
+        }
+      }
+    }
+    return tokens;
+  }
+
+  Node parseIntoTree() {
+    final List<Node> tokens = lexIntoTokens();
+    final List<ST> parsingStack = <ST>[ST.message];
+    final Node syntaxTree = Node(ST.empty, 0, expectedSymbolCount: 1);
+    final List<Node> treeTraversalStack = <Node>[syntaxTree];
+
+    // Helper function for parsing and constructing tree.
+    void parseAndConstructNode(ST nonterminal, int ruleIndex) {
+      final Node parent = treeTraversalStack.last;
+      final List<ST> grammarRule = grammar[nonterminal]![ruleIndex];
+      final Node node = Node(nonterminal, tokens.isNotEmpty ? tokens[0].positionInMessage : -1, expectedSymbolCount: grammarRule.length);
```

Comment:
This is a little more readable:
```suggestion
      final int positionInMessage = tokens.isNotEmpty ? tokens.first.positionInMessage : -1;
      final Node node = Node(nonterminal, positionInMessage, expectedSymbolCount: grammarRule.length);
```

Also can you add a comment explaining what -1 means?

Reply:
I made the change you suggested and added a comment as well.